enhancements/ingress/haproxy-router-compression.md

```diff
@@ -0,0 +1,421 @@
+
+{{- /* compressionMimeTypes stores valid options for HTTP compression MIME Types */}}
+{{- $compressionMimeTypes := "text/html text/css text/plain text/xml text/x-component text/javascript \
+application/x-javascript application/javascript application/json application/manifest+json application/vnd.api+json \
+application/xml application/xhtml+xml application/rss+xml application/atom+xml application/vnd.ms-fontobject \
+application/x-font-ttf application/x-font-opentype application/x-font-truetype \
+image/svg+xml image/x-icon image/vnd.microsoft.icon font/ttf font/eot font/otf font/opentype" -}}
+...
+defaults
+...
+{{- with  $compressionAlg := firstMatch $compressionAlgoPattern {{ env "ROUTER_COMPRESSION_ALGORITHM"}} -}}
+compression algo {{ $compressionAlg }}
+{{- with $compressionMimeTypes := allMatches $compressionMimeTypes {{ env "ROUTER_COMPRESSION_MIME"}} -}}
+compression type {{ $compressMimeTypes }}
+{{- end }}
+{{- end -}}
+```
+The new function `allMatches` will be added to the `template_helper.go` code:
+```go
+// Compare two white-space delimited strings and return the latter if all its substrings
+// exist in former
+func allMatches(defined string, supplied string) string {
+	log.V(7).Info("allMatches called", "supplied", supplied, "defined", defined)
+	numFound := 0
+	suppliedRange := strings.Fields(supplied)
+	definedRange := strings.Fields(defined)
+	for _, s := range suppliedRange {
+		for _, d := range definedRange {
+            if d == s { // optimize this
+                log.V(7).Info("allMatches validated", "supplied", s)
+                numFound++
+                break
+            }
+        }
+	}
+	if numFound == len(suppliedRange){
+        return supplied
+    }
+	return ""
+}
+```
+
+The proposed change to IngressControllerSpec exposes the HTTP compression configuration variables at the default level:
+```go
+type IngressControllerSpec struct {
+...
+    // httpCompression defines policy for HTTP traffic compression
+    // 
+    // The default value is no compression
+    //
+    // +optional
+    HTTPCompression *HTTPCompressionPolicy `json:"httpCompression,omitempty"`
+}
+...
+// httpCompressionPolicy defines the compression algorithm and MIME types that need to
+// have compression applied.
+//
+// This field is optional, and its absence implies that compression should not be enabled
+// globally in HAProxy.
+//
+// If httpCompressionPolicy exists, compression should be enabled only for the listed
+// MIME types, and only using the designated compression algorithm.
+type HTTPCompressionPolicy struct {
+
+	// compressionAlgorithm defines the desired compression algorithm.
+	//
+	// +required
+	CompressionAlgorithm CompressionAlgorithmType `json:"compressionAlgorithm,omitempty"`
+	
+	// compressionMIMETypes is a list of MIME types that should have compression applied
+	//
+	// +kubebuilder:MinItems=1
+	// +required
+	CompressionMIMETypes []string `json:"compressionMIMETypes,omitempty"`
```

Comment:
My lack of http knowledge shows here, but is this possible to make an enum to avoid accidents?

Reply:
Based on feedback from the team, we will not be adding any compressionMIMEType validation or enforcing policy, except maybe a basic regex.

Reply:
Now I have an open question about whether my latest solution is too strict.